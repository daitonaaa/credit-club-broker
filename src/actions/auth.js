import api from 'api';
import url from 'routes/urls';
import { getUser } from 'actions/user';
import { push } from 'connected-react-router';
import * as actionTypes from 'constants/actionTypes';
import { token, take, phoneUtils } from '@creditclub/helpers';


export const authSetField = (field, value) => ({
  type: actionTypes.AUTH_SET_FIELD,
  field, value,
});


export const resetAuth = () => ({
  type: actionTypes.AUTH_RESET,
});


export const login = () => (dispatch, getState) => {
  const { username, password } = getState().auth;
  dispatch(authSetField('fetching', true));

  return api.auth.getToken(phoneUtils.unParse(username), password)
    .then(({ data }) => {
      dispatch(authSetField('success', true));

      setTimeout(() => {
        token.set(data);
        dispatch(push(url.application.path));
      }, 300);
    })
    .catch((e) => {
      const err = take(e, 'response.data', {});
      dispatch(authSetField('fetching', false));

      /** @namespace err.error_description */
      dispatch(authSetField(
        'errorText',
        err.error_description === 'Bad credentials'
          ? 'Не верный логин или пароль'
          : err.error_description
      ))
    });
};

export const logout = () => (dispatch) => {
  token.clear();
  dispatch(push(url.login.path));
};

export const updateToken = (withUpdateUser) => (dispatch) => {
  return api.auth.updateToken()
    .then(({ data }) => {
      token.set(data);

      if (withUpdateUser) {
        dispatch(getUser());
      }

      return data;
    })
    .catch(() => dispatch(logout()));
};
