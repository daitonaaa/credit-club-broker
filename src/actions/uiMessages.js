import uuidv4 from 'uuid/v4';
import * as actionTypes from 'constants/actionTypes';


export const deleteMessage = (id) => ({
  type: actionTypes.UI_MESSAGES_REMOVE_ONE,
  id,
});


export const resetUiMessages = () => ({
  type: actionTypes.UI_MESSAGES_RESET,
});


const addMessage = (event) => ({
  type: actionTypes.UI_MESSAGES_ADD_MESSAGE,
  event,
});


/**
 *
 * @param { string } event.text
 * @param { warn | error | static } event.type
 *
 * @param {object=} event.button
 * @package {string=} event.button.text
 * @param {function=} event.button.handler
 * Если передана функция то отобразит кнопку
 * а при нажатии вызовет её и закроет сообщение
 */
export const addUiMessage = (event) => (dispatch) => {
  if (!event.text) {
    return;
  }

  dispatch(addMessage(
    new UiMessage(event.text, event.type, event.button)
  ));
};

// helpers
export class UiMessage {
  constructor(text, type = 'static', button = null) {
    this.id = uuidv4();
    this.type = type;
    this.text = text;
    this.button = button;
  }
}
