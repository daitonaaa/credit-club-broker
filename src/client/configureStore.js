import logger from 'redux-logger';
import rootReducer from 'reducers';
import thunkMiddleware from 'redux-thunk';
import {createStore, applyMiddleware} from 'redux';
import {routerMiddleware} from 'connected-react-router';

const isProductionBuild = process.env.NODE_ENV === 'production';

const getMiddlewares = (browserHistory) => {
  if (isProductionBuild) return applyMiddleware(
    routerMiddleware(browserHistory),
    thunkMiddleware
  );

  return applyMiddleware(
    routerMiddleware(browserHistory),
    thunkMiddleware,
    logger,
  );
};

// Формируем главное хранилище
export default function configureStore(initialState, browserHistory) {
  return createStore(
    rootReducer(browserHistory),
    initialState,
    getMiddlewares(browserHistory),
  );
}
