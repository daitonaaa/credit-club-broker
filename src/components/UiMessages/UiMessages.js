import cx from 'classnames';
import PropTypes from 'prop-types';
import { isEmpty } from 'ramda';
import { connect } from 'react-redux';
import React, { Component, Fragment } from 'react';

import {
  deleteMessage,
  resetUiMessages,
} from 'actions/uiMessages';

import Message from './Message';

import styles from './UiMessages.module.scss';


class UiMessages extends Component {

  static propTypes = {
    messages: PropTypes.array.isRequired,

    deleteMessage: PropTypes.func.isRequired,
    resetUiMessages: PropTypes.func.isRequired,
  };

  handleDelete = (id) => {
    const { deleteMessage } = this.props;

    deleteMessage(id);
  };

  render() {
    const {
      messages,
      resetUiMessages,
    } = this.props;

    return (
      <div className={cx(styles.wrapper, {
        [styles.hasMessages]: Boolean(messages.length),
      })}>
        {
          !isEmpty(messages) && (
            <Fragment>
              {
                messages.length > 1 && (
                  <div
                    onClick={resetUiMessages}
                    className={styles.closeall}
                  >
                    Скрыть
                  </div>
                )
              }
              {
                messages.map(item => (
                  <Message
                    {...item}
                    key={item.id}
                    onDelete={this.handleDelete}
                  />
                ))
              }
            </Fragment>
          )
        }
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  ...state.uiMessages,
});

const mapDispatchToProps = {
  deleteMessage,
  resetUiMessages,
};

export default connect(mapStateToProps, mapDispatchToProps)(UiMessages);
