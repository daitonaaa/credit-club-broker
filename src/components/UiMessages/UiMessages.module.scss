@import '~vars', '~mixins';

.wrapper {
  position: fixed;
  z-index: 999;
  overflow-y: auto;
  top: 60px;
  right: 0;
  margin-bottom: 25px;
  padding: 20px;
  max-height: calc(100% - 90px);
  transform: translateX(40px);
  transition: $baseTransition;
}

.has-messages {
  transform: translateX(0);
}

.closeall {
  @include flexBox(center);
  margin-bottom: 8px;
  margin-left: calc(100% - 85px);
  width: 85px;
  height: 30px;
  background: #fff;
  color: $blue;
  border-radius: $border;
  box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.08), 0 1px 2px 0 rgba(0, 0, 0, 0.09);

  &:hover { cursor: pointer; }
}
