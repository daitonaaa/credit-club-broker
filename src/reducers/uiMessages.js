import Immutable from 'immutable';

import { immutableize } from '@creditclub/helpers';

import * as actionTypes from 'constants/actionTypes';


const initialState = Immutable.fromJS({
  messages: [],
});


const uiMessages = (state = initialState, action) => {
  switch (action.type) {

    case actionTypes.UI_MESSAGES_ADD_MESSAGE:
      return state.update('messages', list => list.push(action.event));

    case actionTypes.UI_MESSAGES_REMOVE_ONE: {
      const itemIndex = state.get('messages').findIndex(
        item => item.id === action.id
      );

      return state.deleteIn(['messages', itemIndex]);
    }

    case actionTypes.UI_MESSAGES_RESET:
      return initialState;

    default:
      return state;
  }
};


export default immutableize(uiMessages);
