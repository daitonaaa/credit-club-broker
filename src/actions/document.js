import co from 'co';
import api from 'api';
import { isEmpty } from 'ramda';
import * as Sentry from '@sentry/browser';
import { addError } from 'actions/error';
import * as actionTypes from 'constants/actionTypes';


const updateFiles = (files) => ({
  type: actionTypes.DOCUMENT_UPDATE_FILES,
  files,
});


export const deleteFile = (fileName) => ({
  type: actionTypes.DOCUMENT_DELETE_FILE,
  fileName,
});


export const resetDocument = () => ({
  type: actionTypes.DOCUMENT_RESET,
});


const updateFileToProgress = (fileName, progress) => ({
  type: actionTypes.DOCUMENT_UPDATE_FILE_UPLOAD_PROGRESS,
  fileName, progress,
});


const uploadProgress = ({ loaded, total }, file) => (dispatch) => {
  const progress = Math.round(loaded * 100 / total);

  dispatch(updateFileToProgress(file.name, progress));
};


export const onDropFiles = (files, unsupportedFiles) => (dispatch, getState) => {
  if (!isEmpty(unsupportedFiles)) {
    unsupportedFiles.forEach((file) => console.error(`file ${file.name} is not supported`))
  }

  if (!isEmpty(files)) {
    const filesList = [];
    const { files: attach } = getState().document;

    files.forEach((file) => {
      if (
        file.size > 0 &&
        !attach.find(({ name }) => name === file.name)
      ) {
        filesList.push(file);
      }
    });

    dispatch(updateFiles(filesList));
  }
};


export const sendFiles = () => async (dispatch, getState) => {
  const { files } = getState().document;
  if (isEmpty(files)) return null;

  const { data } = await api.document.getTemporaryUrls((
    files.map(file => file.name)
  ));

  Sentry.captureEvent({
    message: `Брокер ${getState().user.data.fullName} отправил файлы`,
    level: 'log',
    extra: {
      startUploadDate: new Date().toString(),
      files: {
        totalCount: files.length,
        data: files.map(({ name, size, type }) => ({
          name, size, type,
          temporary: data.find(item => item.filename === name),
        })),
      }
    }
  });

  function* sendFilesGen() {
    for (let i = 0; i < data.length; i++) {
      const dataFile = data[i];
      const file = files.find(item => item && item.name === dataFile.filename);

      yield api.document
        .sendFile(dataFile.url, file, file.type, (progress) => dispatch(uploadProgress(progress, file)))
        .catch((err) => {
          dispatch(addError({
            err: {
              origin: err,
              extra: {
                actionMethod: 'api.document.sendFile',
              }
            }
          }));
        });
    }
  }

  await co(sendFilesGen());

  return data.map(item => item.objectKey);
};
