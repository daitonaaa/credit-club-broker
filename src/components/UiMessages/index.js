import UiMessages from './UiMessages';

export default UiMessages;
