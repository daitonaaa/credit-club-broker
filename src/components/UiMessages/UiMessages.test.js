import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { UiMessage } from 'actions/uiMessages';

import UiMessages from './UiMessages';


describe('UiMessages', () => {
  const mockStore = configureStore();
  const initialState = {
    uiMessages: {
      messages: [],
    }
  };
  let store, container;

  beforeEach(() => {
    store = mockStore(initialState);
    container = mount(<UiMessages store={store} />);
  });

  test('Рендер компонента', () => {
    console.log(container.debug());
    expect(container).toMatchSnapshot();
  });
});
