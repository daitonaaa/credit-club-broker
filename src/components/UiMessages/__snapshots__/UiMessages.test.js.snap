// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UiMessages Рендер компонента 1`] = `
<Connect(UiMessages)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <UiMessages
    deleteMessage={[Function]}
    messages={Array []}
    resetUiMessages={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <div
      className="wrapper"
    />
  </UiMessages>
</Connect(UiMessages)>
`;
